### ISSUE TITLE
nlmixr2() no longer returning dataframe

### ISSUE THREAD

**Issue by DrOJG** [nlmixr2/nlmixr2#235]
Hello there,

A piece of model code that was previously working for estimation with nlmixr2() is now no longer returning the data frame portion of the model fit object.

While estimation still works, and other portions of the model fit are recoverable (OFV, parameter estimates etc.), the main data frame is not there. Under run info, I see it now states "error calculating tables, returning without table step".

This seemed to coincide with some system updates, so I am wondering if there are any incompatibilities with new versions of compilers or other dependencies I should be aware of?

Many thanks,
O.

**Comment by billdenney** [comment]
I don't think that there are any known version incompatibilities.  But, changing libraries may possibly require recompiling the nlmixr2 tools because of linking changes.  Can you try reinstalling nlmixr2 and see if it works then?

**Comment by DrOJG** [comment]
Hi,
After uninstalling nlmixr2 and all dependencies, purging the renv cache, and reinstalling, the problem is still persisting.
Happy to provide any info that might help with solving.
Best,
O.

**Comment by mattfidler** [comment]
This is likely not the case for all models.  If there is an error in the table step, nlmixr will return an object without the table step.

If the data are not propritary and the model is not propriatary the easiest way to provide a reproducible example is to save the model and then use that to try to add tables again.  I believe the function is `calcTable()` but I dont currently have access to a computer so this is based on memory.

Once I have a reproducible example I can try to figure out what is going on in your case.

**Comment by DrOJG** [comment]
Hi Matt,
At this point I would prefer not to share the dataset or the complete model as we are looking to publish soon, I apologise for this.
I have however done some further investigation which will hopefully be useful in diagnosing.

I am currently running the model fitting with the below command:
```
model.fit <- nlmixr2(model, data_set, est = "foce", table = tableControl(cwres = TRUE, npde = TRUE))
```
As stated previously, this is returning a fit object without the table step. When I try to add a table to the object with the `addTable()`, it throws the following error:
```
→ Calculating residuals/tables
[====|====|====|====|====|====|====|====|====|====] 0:00:00 

Error in `$<-.data.frame`(`*tmp*`, "sim.id", value = numeric(0)) : 
  replacement has 0 rows, data has 302
In addition: Warning message:
In max(.sim$sim.id) : no non-missing arguments to max; returning -Inf
```
Additionally, I have tried running both the Friberg myelosuppression example model and the Nimotuzumab example model in the same environment and both returned model fit dataframes. It is clearly a specific issue, but strange that this code previously ran fine with no alterations to the model or dataset.

Thanks,
O.

**Comment by mattfidler** [comment]
If you can add a traceback it would be more helpful.  Currently I have insufficient information to do anything about this ☹️

**Comment by mattfidler** [comment]
This is done with `traceback()` from the place after `addTable()`.  This way it gives where the error occurs without providing the dataset or model.

**Comment by DrOJG** [comment]
Sorry Matt, I have been on leave so was not able to respond!
The following command:
```
model_with_table <- addTable(model.fit.object, table = tableControl(cwres = TRUE, npde = TRUE))
```
throws the aforementioned error, the `traceback()` is as follows:
```
> traceback()
11: stop(sprintf(ngettext(N, "replacement has %d row, data has %d", 
        "replacement has %d rows, data has %d"), N, nrows), domain = NA)
10: `$<-.data.frame`(`*tmp*`, "sim.id", value = numeric(0))
9: `$<-`(`*tmp*`, "sim.id", value = numeric(0))
8: vpcSim(fit, n = table$nsim, seed = table$seed, addDosing = addDosing, 
       subsetNonmem = subsetNonmem)
7: .calcCwres0(fit, data, thetaEtaParameters, table, dv = dv, predOnly, 
       addDosing, subsetNonmem, keep, npde, .prdLst = .prdLst)
6: .calcCwres(fit, data = fit$dataSav, thetaEtaParameters = .thetaEtaParameters, 
       table = table, dv = .ret[[1]][[1]], predOnly = .predOnly, 
       addDosing = table$addDosing, subsetNonmem = table$subsetNonmem, 
       keep = keep, .prdLst = .prdLst, npde = .npde2)
5: .calcTables(.fit, data = data, table = table, keep = keep)
4: force(code)
3: nlmixrWithTiming("table", {
       keep <- unique(c(keep, "nlmixrRowNums"))
       .malert("Calculating residuals/tables")
       .objName <- substitute(object)
       if (!inherits(object, "nlmixr2FitCore")) {
           stop("requires a nlmixr2 fit object", call. = FALSE)
       }
       .fit <- object$env
       if (exists("origControl", .fit)) {
           .control <- .fit$origControl
       }
       else if (exists("control", .fit)) {
           .control <- .fit$control
       }
       else {
           .control <- foceiControl()
       }
       if (is.null(.fit$omega)) {
           .df <- .calcIres(.fit, data = data, table = table, dv = NULL, 
               addDosing = table$addDosing, subsetNonmem = table$subsetNonmem, 
    ...
2: nlmixr2est::addTable(object = object, updateObject = updateObject, 
       data = data, thetaEtaParameters = thetaEtaParameters, table = table, 
       keep = keep, drop = drop, envir = envir)
1: addTable(model.fit.object, table = tableControl(cwres = TRUE, 
       npde = TRUE))
```
Let me know if this gives any insight. I am also conducting my own investigations into this to narrow down what specifically is causing the error.

p.s. Thank you for your efforts here, I appreaciate it!

**Comment by DrOJG** [comment]
Hi Matt,
To update further, I appear to have isolated the issue. 

A single covariate, coded in the form `tL0 + beta_pbcb_L0 * PB_or_CB + eta.L0` where `PB_or_CB` is a binary indicator variable in the dataset, was causing the issue. Removing specifically this covariate from the model fixed the issue, albeit changing the fit slightly.

I am still not sure why this was causing an issue, as it was previously included without causing problems, but it does appear to be the source of the error. It may be it was causing some instability in the fit but I am not sure what changed to cause it to break the table calculation. 

Thanks,
O.

**Comment by billdenney** [comment]
That is very helpful diagnosis.  Is the covariate column always or almost always the same value?

**Comment by DrOJG** [comment]
Hi Bill,
As I say, it is a simple binary covariate (`0` or `1`), but I have just checked through the dataset and found it is set to `1` in approximately 20% of rows. 
The parameter estimate when included was approx. `-0.259`, but has a fairly wide confidence interval of `(-0.481, -0.0373)`
Thanks,
O.

**Comment by mattfidler** [comment]
Hi @DrOJG 

I believe that this happens with the `vpcSim()` function that simulates both the vpc and the information for `npde`.  As a confirmation you remove `npde` it is likely that this will return a data frame.

The problem is with this covariate (and its large range), sometimes it selects models that cannot be solved.  In general, these simulations are rejected and then another simulation replaces the simulations that could not be solved.

You have reached an edge case where this replacement isn't working well.  I will try to help you debug this further.  First I will transfer this to `nlmixr2est` where the `vpcSim()` model lives.
